### PR TITLE
Update doc for finetuner

### LINKED ISSUE #189: Update documentation for finetuner
#### Problem description
Require to update finetuner doc

### PR DESCRIPTION
**Pull request by YuanmingLeee**
Closes #189 

#### What has been done to verify that this works as intended?
N.A., doc related.

#### Why is this the best possible solution? Were any other approaches considered?
N.A.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
N.A.

#### Does this change require updates to documentation?
Yes.

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/app/v1/endpoints/model.py [hunk 1]
```diff
@@ -48,12 +49,30 @@ async def get_model_structure(id: str):  # noqa
 
 
 @router.patch('/structure/{id}')  # TODO: add response_model
-def update_model_structure_as_new(id: str, dry_run: bool = False):  # noqa
+def update_model_structure_as_new(id: str, structure, dry_run: bool = False):  # noqa
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/model.py:52]: how about changing the `structure` to `updated_layer`?

File: modelci/app/v1/endpoints/trainer.py [hunk 2]
```diff
@@ -13,13 +13,13 @@
...
-    Submit a training job.
+    Submit a training job to the controller.
 
     Args:
         model_id (str): Model ID for training.
         other_body: Other required parameters for training.
 
     Returns:
-        Job data class object.
+        Submitted training job data class object.
```
Comment (HuaizhengZhang) [inline modelci/app/v1/endpoints/trainer.py:23]: where is the training job id generator?

Better to make this design and comments more clear

File: modelci/finetuner/README.md [hunk 3]
```diff
@@ -1,7 +1,84 @@
...
-Tune ResNet18 on CIFAR10 for 25 epochs:
-```shell script
-python computer_vision_fune_tuning.py
+\[Experimental\] Finetuner is a tiny module of ModelCI for easy fine-tuning pre-trained model on a new dataset. 
+
+Finetuner **packs the training job** with specified Transfer Learning algorithm, dataset, and pre-trained model 
+written on various ML framework (e.g. PyTorch). Such training jobs are **managed and schedules** by a coordinator. 
+We are to support various training environment, such as local machine, private cluster, and public cloud platform. 
+
+We leverage [PyTorch Lihgtning](https://github.com/PyTorchLightning/pytorch-lightning) as the training backend for
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:9]: do not need this pytorch lightning introduction

File: modelci/finetuner/README.md [hunk 4]
```diff
@@ -1,7 +1,84 @@
...
-## Fine tune for Computer Vision
-Tune ResNet18 on CIFAR10 for 25 epochs:
-```shell script
-python computer_vision_fune_tuning.py
+\[Experimental\] Finetuner is a tiny module of ModelCI for easy fine-tuning pre-trained model on a new dataset. 
+
+Finetuner **packs the training job** with specified Transfer Learning algorithm, dataset, and pre-trained model 
+written on various ML framework (e.g. PyTorch). Such training jobs are **managed and schedules** by a coordinator. 
+We are to support various training environment, such as local machine, private cluster, and public cloud platform. 
+
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:8]: Besides, we plan to adopt auto tuning tools such as NNI and autoKeras into the system to free users from heavy training hyper-parameter tunining tasks.

File: modelci/finetuner/README.md [hunk 5]
```diff
@@ -1,7 +1,84 @@
...
+We leverage [PyTorch Lihgtning](https://github.com/PyTorchLightning/pytorch-lightning) as the training backend for
+PyTorch models.
+
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:11]: ## Workflow
- [ ] todo

File: modelci/finetuner/README.md [hunk 6]
```diff
@@ -1,7 +1,84 @@
...
+                </ul>
+            </td>
+        </tr>
+    </tbody>
+</table>
+
+## Tutorials
+
+1. Fine tune a ResNet18
+2. Fine tune a BERT-base-cased
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:84]: ## Roadmap

- [ ] Add the workflow figure
- [ ] Deep-Q Learning support
- [ ] NNi support
- [ ] ...
......

File: modelci/finetuner/README.md [hunk 7]
```diff
@@ -1,7 +1,84 @@
...
+written on various ML framework (e.g. PyTorch). Such training jobs are **managed and schedules** by a coordinator. 
+We are to support various training environment, such as local machine, private cluster, and public cloud platform. 
+
+We leverage [PyTorch Lihgtning](https://github.com/PyTorchLightning/pytorch-lightning) as the training backend for
+PyTorch models.
+
+## Structure
+```text
+├── coordinator           # manages and schedules training jobs
+├── torch_data_module.py  # creates data loader from dataset
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:15]: pytorch_dataloader.py

File: modelci/finetuner/README.md [hunk 8]
```diff
@@ -1,7 +1,84 @@
...
+├── tl_module.py          # defines transfer learning algorithm
```
Comment (HuaizhengZhang) [inline modelci/finetuner/README.md:16]: transfer_learning.py

**Review by HuaizhengZhang — COMMENTED** [review]
Better than before

File: modelci/finetuner/README.md (hunk 7 above)
Comment (YuanmingLeee) [reply modelci/finetuner/README.md]: `pytorch_datamodule` is better. It is named from `torch_lightning.LightningDataModule`.

File: modelci/app/v1/endpoints/model.py (hunk 1 above)
Comment (univerone) [reply modelci/app/v1/endpoints/model.py:52]: How about model_operations?

File: modelci/app/v1/endpoints/model.py [hunk 9]
```diff
@@ -48,16 +49,81 @@ async def get_model_structure(id: str):  # noqa
...
+            [D] fc
+            [A] fc1: (nn.Linear) in_features=1024, out_features=512
+            [A] fc2: (nn.Linear) in_features=512, out_features=10
+        Besides, we have connection changes:
+            [D] conv1 -> fc
+            [A] conv1 -> fc1
+            [A] fc1 -> fc2
+
+        >>>
+        ... structure = {
```
Comment (univerone) [inline modelci/app/v1/endpoints/model.py:81]: How about change this structure as 
```
'layer': [
{ layer_name: 'fc', '_op': 'D' },
{ layer_name: 'fc1', '_op': 'A' , 'in_features': 1024, 'out_features': 512, '_type': 'nn.Linear' }
],
'connection': [
{'in_layer': 'fc', 'out_layer': 'fc1', '_op': 'A'}
]
```

File: modelci/app/v1/endpoints/model.py (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:52]: I create a temp new function `update_finetune_model_as_new`. Currently, we use this function to do model fine-tune update.

File: modelci/app/v1/endpoints/model.py (hunk 9 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:81]: This increases redundancy. What is the benefit if we switch to such a structure?

Comment (univerone) [reply modelci/app/v1/endpoints/model.py:81]: > This increases redundancy. What is the benefit if we switch to such a structure?

Yeah, it's maybe more complicated. How about implement a sorted dict data strcuture as https://github.com/martinjm97/ENNUI/blob/49beaee37579fd094b11f87cad6d86a6c549105b/src/ui/shapes/layer.ts#L25, we only return an ordered dict for each model, and recreate the whole model based on that as modifying old model can be a tricky job.

Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/model.py:81]: 1. Yes I agree with the idea of using an ordered dict for the `layer` part of the structure.

2. In my view, recreate the whole model based on the new structure is more complicated. 
ENNUI (the link you provided) does not support training from a pre-trained model.  Therefore, when there is an update, it retrains from scratch.
We are supposed to support start from a pre-trained one. We need to compare the new full version and the old full version, see which layers get updated, and retain those layers that are not changed. Let the frontend to keep track of the changed layers by the user is a much easier job.

File: modelci/app/v1/endpoints/trainer.py (hunk 2 above)
Comment (YuanmingLeee) [reply modelci/app/v1/endpoints/trainer.py:24]: Updated doc.

**Review by univerone — APPROVED** [review]
LGTM

**Pull request merged**